rs/sns/governance/src/types.rs

```diff
@@ -2160,6 +2164,236 @@ impl NeuronParameters {
+                        }
+                        Err(e) => {
+                            defects.push(e);
+                        }
+                    }
+                } else {
+                    defects.push(
+                        "Missing nns_neuron_hotkeys for neurons fund participant".to_string(),
+                    );
+                }
+            }
+            None => {
+                defects.push("Missing participant type (Direct or Neurons' Fund)".to_string());
+            }
+        }
+
+        if !defects.is_empty() {
+            let participant_info = match participant {
+                Some(Participant::Direct(_)) => {
+                    format!("direct participant {:?}", self.controller)
+                }
+                Some(Participant::NeuronsFund(nf)) => {
+                    format!("neurons fund participant {:?}", nf.nns_neuron_id)
+                }
+                None => "unknown participant".to_string(),
+            };
+
+            Err(format!(
+                "Could not claim neuron for {} with NeuronId {:?} due to: {}",
+                participant_info,
+                neuron_id,
+                defects.join("\n"),
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    pub(crate) fn is_community_fund_neuron(&self) -> bool {
+        matches!(self.participant, Some(Participant::NeuronsFund(_)))
+    }
+
+    pub(crate) fn get_controller_or_panic(&self) -> PrincipalId {
+        self.controller
+            .expect("Expected the principal to be present for neuron")
+    }
+
+    pub(crate) fn get_dissolve_delay_seconds_or_panic(&self) -> u64 {
+        self.dissolve_delay_seconds
+            .expect("Expected the dissolve_delay_seconds to be present in NeuronRecipe")
+    }
+
+    pub(crate) fn get_stake_e8s_or_panic(&self) -> u64 {
+        self.stake_e8s
+            .expect("Expected the stake_e8s to be present in NeuronRecipe")
+    }
+
+    pub(crate) fn get_neuron_id_or_panic(&self) -> &NeuronId {
+        self.neuron_id
+            .as_ref()
+            .expect("Expected NeuronId to be present in NeuronRecipe")
+    }
+
+    pub(crate) fn source_nns_neuron_id(&self) -> Option<u64> {
+        match &self.participant {
+            Some(Participant::NeuronsFund(nf)) => nf.nns_neuron_id.as_ref().cloned(),
+            _ => None,
+        }
+    }
+
+    pub(crate) fn construct_permissions_or_panic(
+        &self,
+        neuron_claimer_permissions: NeuronPermissionList,
+    ) -> Vec<NeuronPermission> {
+        self.construct_permissions(neuron_claimer_permissions)
+            .expect("Failed to construct permissions for neuron")
+    }
+
+    pub(crate) fn construct_permissions(
+        &self,
+        neuron_claimer_permissions: NeuronPermissionList,
+    ) -> Result<Vec<NeuronPermission>, String> {
+        let mut permissions = vec![];
+
+        permissions.push(NeuronPermission::new(
+            &self.get_controller_or_panic(),
```

Comment:
in the case of NeuronsFund neurons, will the controller here be NNS Governance?

Reply:
Yes it will be, this field is decided by swap